xarray/core/rolling_exp.py

```diff
@@ -1,8 +1,17 @@
+from typing import TYPE_CHECKING, Generic, Hashable, Mapping, Optional, TypeVar
+
 import numpy as np
 
+from .options import _get_keep_attrs
 from .pdcompat import count_not_none
 from .pycompat import is_duck_dask_array
 
+if TYPE_CHECKING:
+    from .dataarray import DataArray  # noqa: F401
+    from .dataset import Dataset  # noqa: F401
+
+T_DSorDA = TypeVar("T_DSorDA", "DataArray", "Dataset")
```

Comment:
Is this different from `DataWithCoords`?

Reply:
```python
from .common import DataWithCoords
T_DataWithCoords = TypeVar("T_DataWithCoords", bound=DataWithCoords)
```
works as well. A bit to my surprise as `DataWithCoords` does not implement `reduce`. But mypy does not throw an error here: 

```python
from xarray.core.common import DataWithCoords
class A:
    pass

hasattr(xr.core.common.DataWithCoords, "reduce") # -> False

def test(x: "A"):
    x.reduce() # mypy errors

def test2(x: "DataWithCoords"):
    x.reduce() # mypy does not error
```
not sure why not...

---

(Also I am not entirely sure what the difference is between
```python
T_Parent = TypeVar("T_Parent", bound=Parent)
T_Children = TypeVar("T_Children", ChildA, ChildB)
```
)

Reply:
That is odd it doesn't throw an error... 

On the second case, I think — not completely sure — that `T_Parent` is generic with an "upper bound" of `Parent`, so any function that takes `T_Parent` will return the same type. 

Whereas a function that takes `T_Children` could return either of `ChildA` or `ChildB` regardless of what's passed in.

Reply:
And your call on whether we use `DataWithCoords` or the `TypeVar`, but I think if we do the latter we could find a clearer name :)

I would marginally vote to go with `DataWithCoords` given we do that elsewhere but totally open on trying another approach

Reply:
I think both would need to be `TypeVar`s. I I'll stick to `T_DSorDA` as `DataWithCoords` does not implement `reduce`. `T_DSorDA` is already used at some places (`map_blocks`), e.g.:

https://github.com/pydata/xarray/blob/5296ed18272a856d478fbbb3d3253205508d1c2d/xarray/core/parallel.py#L34

Reply:
Ah I didn't know that, thanks. I wasn't a fan of the name, but we can change them together another time if others agree.